base/server/src/com/netscape/cms/realm/PKIRealm.java

```diff
@@ -73,10 +75,21 @@ public Principal authenticate(String username, String password) {
 
             return getPrincipal(username, authToken);
 
-        } catch (Exception e) {
+        } catch (EMissingCredential | EInvalidCredentials e) { // authentication failure
 
             logger.warn("Unable to authenticate user " + username + ": " + e.getMessage());
 
+            signedAuditLogger.log(AuthEvent.createFailureEvent(
+                        auditSubjectID,
+                        IAuthSubsystem.PASSWDUSERDB_AUTHMGR_ID,
+                        attemptedAuditUID));
+
+            throw null;
```

Comment:
I believe as per the attached tomcat documentation, we should be `return` null and not `throw` null

Reply:
Ah yes, fixing..